packages/protocol/contracts-0.8/common/MintGoldSchedule.sol

```diff
@@ -0,0 +1,313 @@
+  uint256 public l2StartTime;
+  uint256 public totalSupplyAtL2Start;
+
+  uint256 public totalMintedBySchedule;
+  address public communityRewardFund;
+  address public carbonOffsettingPartner;
+
+  FixidityLib.Fraction private communityRewardFraction;
+  FixidityLib.Fraction private carbonOffsettingFraction;
+
+  event CommunityRewardFractionSet(uint256 fraction);
+  event CarbonOffsettingFundSet(address indexed partner, uint256 fraction);
+
+  modifier whenActivated() {
+    require(areDependenciesSet, "Minting schedule has not been activated.");
+    _;
+  }
+
+  /**
+   * @notice Sets initialized == true on implementation contracts
+   * @param test Set to true to skip implementation initialization
+   */
+  constructor(bool test) public Initializable(test) {}
+
+  /**
+   * @notice A constructor for initialising a new instance of a MintGoldSchedule contract.
+   */
+  function initialize() external initializer {
+    _transferOwnership(msg.sender);
+  }
+
+  /**
+   * @notice Sets the minting schedule dependencies during L2 transition.
+   * @param _l2StartTime The timestamp of L1 to L2 transition.
+   * @param _communityRewardFraction The percentage of rewards that go the community funds.
+   * @param _carbonOffsettingPartner The address of the carbon offsetting partner.
+   * @param _carbonOffsettingFraction The percentage of rewards going to carbon offsetting partner.
+   * @param registryAddress Address of the deployed contracts registry.
+   */
+  function activate(
+    uint256 _l2StartTime,
+    uint256 _communityRewardFraction,
+    address _carbonOffsettingPartner,
+    uint256 _carbonOffsettingFraction,
+    address registryAddress
+  ) external onlyOwner onlyL2 {
+    require(!areDependenciesSet, "Contract has already been activated.");
+    require(registryAddress != address(0), "The registry address cannot be the zero address");
+    require(block.timestamp > _l2StartTime, "L2 start time cannot be set to a future date.");
+    areDependenciesSet = true;
+    l2StartTime = _l2StartTime;
+    setRegistry(registryAddress);
+    communityRewardFund = address(getGovernance());
+    totalSupplyAtL2Start = getGoldToken().totalSupply();
+    setCommunityRewardFraction(_communityRewardFraction);
+    setCarbonOffsettingFund(_carbonOffsettingPartner, _carbonOffsettingFraction);
+  }
+
+  /**
+   * @notice Mints CELO to the community and carbon offsetting funds according to the predefined schedule.
+   */
+  function mintAccordingToSchedule() external nonReentrant onlyL2 returns (bool) {
+    (
+      uint256 targetGoldTotalSupply,
+      uint256 communityRewardFundMintAmount,
+      uint256 carbonOffsettingPartnerMintAmount
+    ) = getTargetGoldTotalSupply();
+
+    uint256 mintableAmount = targetGoldTotalSupply - getGoldToken().totalSupply();
+
+    require(mintableAmount > 0, "Mintable amount must be greater than zero");
+    require(
+      getRemainingBalanceToMint() >= mintableAmount,
```

Comment:
Should we rather mint whatever is possible ?

mintableAmount = Math.min(getRemainingBalanceToMint(), targetGoldTotalSupply - getGoldToken().totalSupply())